### ISSUE TITLE
average coverage and coverage plots

### ISSUE THREAD

**Issue by 0xaf1f** [samtools/samtools#596]
For our read mappings, we always check for average coverage and it's nice to have a plot of coverage as a function of position as well. Lots of people have written their own scripts for this, but it would be helpful if this was standardized in `samtools` and `plot-bamstats`, especially since these produce very useful reports that are just missing a couple things like this.

Many thanks and regards

**Comment by tseemann** [comment]
@afif-elghraoui i don't think it is the job of `samtools` to produce plots. I think `samtools depth -aa file.bam` provides all the information you need. Another option is to use the `picard-tools` package.

**Comment by afif-elghraoui** [comment]
Thanks for your comment. The `plot-bamstats` command included in the `samtools` package here already produces some plots based on the output of `samtools stats`. I just thought that a coverage plot would be a good addition to the report that `plot-bamstats` produces.

**Comment by tseemann** [comment]
@afif-elghraoui My apologies - i did not realise `plot-bamstats` and the other (old) scripts were still included in `samtools 1.3` !

**Comment by fbreitwieser** [comment]
I've just written code to calculate and plot the coverage. The code is based on `samtools depth`, so I could either integrate it with that command, or add a new command `samtools coverage`. If that's of interest I'd be happy to contribute it.

Output:
```
 ./bamcov CEbe_67762_S4_merged.Aspergillus-flavus-NRRL3357.bam | head
#CHROM  LENGTH  N_COVERED_BASES PERCENT_COVERED AVG_COVERAGE
NW_002479957.1  514     335     65.1751 2.05642
NW_002477290.1  2754    626     22.7306 0.268337
NW_002479422.1  934     303     32.4411 0.324411
NW_002477246.1  2012514 234753  11.6647 0.147286
NW_002477303.1  2616    454     17.3547 0.265673
NW_002479737.1  847     86      10.1535 0.157025
```

Histogram is plotted with `-H`, and regions may be specified with `-r`. I could also add PNG or SVG plot output.
```
 ./bamcov -H -r NW_002477246.1:5000-100000 CEbe_67762_S4_merged.Aspergillus-flavus-NRRL3357.bam
NW_002477246.1 (2.01Mbp)
>  31.17% │    █                                             │ Number of reads:  168
>  27.71% │    █                                             │
>  24.24% │    █   █  █                                      │ Covered bases:    9.8Kbp
>  20.78% │    █  ██  █          █  █                        │ Percent covered:  0.487%
>  17.32% │    █  ██  █    █     █  █         ██          █  │ Average coverage: 0.00624x
>  13.85% │    █  ███ █ ██ █     █  █     ██  ███ █ █     █  │ Average baseQ:    34.8
>  10.39% │    █  █████ ████     ██ █     ██  █████ █     █  │ Average mapQ:     43.1
>   6.93% │   ██ ██████ ███████  ██ █ █   ██ ██████████ █ ██ │
>   3.46% │  ███ ██████ ███████  ██ █ █   ██ ██████████ █ ██ │ Histo bin width:  1.9Kbp
>   0.00% │  ███ ██████ ███████  ██ █ █   ██ ██████████ █ ██ │ Histo max bin:    34.632%
         5.0K     24.0K     43.0K     62.0K     81.0K      100.0K
```

**Comment by jkbonfield** [comment]
Hello,

This looks good and many thanks for the offer.

We're currently busy on merging various PRs and preparing for a release sometime in the early new year, so it's not likely something to can look into immediately.  However we are indeed interested in it.  Thanks.

How similar to samtools depth is it?  Are we simply talking a few tweaks to the algorithm to allow it to gather data in a histogram and plot it, or a large amount of rewriting?

**Comment by fbreitwieser** [comment]
The basic logic is the same, though it adds a significant amount of code. Where `samtools depth` outputs the position and depth for each base, I increment the number of covered positions in the respective bin. By default it's 50 bins, but that can be changed with an argument to `-w`. `-w 0` uses the full width of the terminal. Whenever a new sequence is seen, a histogram or table line is printed. A couple of auxiliary statistics are recorded in the loop, such as the number of reads, mapQ and baseQ.

**Comment by jkbonfield** [comment]
Feel free to make a pull request, or if you have questions on appropriate style and integration ask them here first.  It may be easier to discuss once we see the PR though.

**Comment by fbreitwieser** [comment]
@jkbonfield , I made the pull request #992. I added the functionality as the new command 'samtools coverage'. Since there are a couple of distinctions from samtools depth, including not supporting BED for now and a different tabular format, I think a new command works better than trying to fit it in with depth. Looking forward to your feedback.

**Comment by valeriuo** [comment]
Fixed by #992
